### PR TITLE
Cluster-autoscaler: cloud provider interface implementation for AWS

### PR DESCRIPTION
**Pull request by osxi**
Part of https://github.com/kubernetes/contrib/issues/1311

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1377)

### CODE REVIEW

**Comment by wojtek-t** [comment]
@mwielgus @piosz @fgrzadkowski

**Comment by osxi** [comment]
I signed it!

**Comment by mwielgus** [comment]
cc:  @andrewsykim @pbitty @iterion

File: cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	AwsRef
+
+	awsManager *AwsManager
+
+	minSize int
+	maxSize int
+}
+
+// MaxSize returns maximum size of the node group.
+func (asg *Asg) MaxSize() int {
```
Comment (iterion) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go]: It seems we could just pull MaxSize and MinSize from the AutoScalingGroup when we initialize. It's kind of annoying that it must be specified twice, both here and in the ASG itself. I'm not sure if there would be a use case for that?

@mwielgus I think this was lifted pretty much verbatim from the GCE side. I assume (from my short search through the MIG docs) that MIGs don't allow you to specify a min/max?

Comment (pbitty) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go]: That makes sense to me.

File: cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	if err != nil {
+		return nil, err
+	}
+	asg, err := aws.awsManager.GetAsgForInstance(ref)
+	return asg, err
+}
+
+// AwsRef contains a reference to some entity in AWS/GKE world.
+type AwsRef struct {
+	Zone string
```
Comment (pbitty) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go]: What does a zone represent?  If it is an availability zone, the autoscaling groups span availability zones, so zone is not part of the group's identity.

Comment (pbitty) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go]: On that note, do we care about region?  I imagine for the initial implementation we can just default to the region that the cluster-autoscaler is running in.  It sounds like the recommended pattern is to limit a cluster to a region anyway.

Thoughts?

File: cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go]: @iterion No, migs allow you only to specify the target number of instances.

**Comment by andrewsykim** [comment]
I signed it!

**Comment by mwielgus** [comment]
@pbitty @iterion Are you ok with the implementation? Did you have a chance to test it?

**Comment by iterion** [comment]
We haven't tested the most recent iteration, but there were only minor changes. We'll try to test it all again soon.

**Comment by andrewsykim** [comment]
Same thing on our side, hopefully this week we'll get a chance to test it again. We're also looking into writing some e2e tests for the AWS implementation.

**Comment by osxi** [comment]
I just tested it out and it works good.

**Comment by mwielgus** [comment]
@osxi could you please sign the CLA?

File: cluster-autoscaler/cloudprovider/aws/aws_cloud_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+// 	provider := testProvider(t)
+// 	err := provider.addNodeGroup("1:5:test-asg")
+// 	assert.NoError(t, err)
+// 	assert.Equal(t, len(provider.asgs), 1)
+// 	targetSize, err := provider.asgs[0].TargetSize()
+// 	assert.Equal(t, targetSize, 1)
+// 	assert.NoError(t, err)
+// }
+
+// TODO: IncreaseSize
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider_test.go]: Please implement.

File: cluster-autoscaler/cloudprovider/aws/aws_cloud_provider_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+func TestMinSize(t *testing.T) {
+	provider := testProvider(t)
+	err := provider.addNodeGroup("1:5:test-asg")
+	assert.NoError(t, err)
+	assert.Equal(t, len(provider.asgs), 1)
+	assert.Equal(t, provider.asgs[0].MinSize(), 1)
+}
+
+// TODO: Mock aws api response
+// func TestTargetSize(t *testing.T) {
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider_test.go]: Uncomment or remove.

**Comment by osxi** [comment]
@mwielgus I signed the CLA but then I added commit from @andrewsykim, who also signed the CLA.

I'll implement the commented tests.

**Comment by andrewsykim** [comment]
Finally had some time today to test this out! Sorry it took so long, we had to upgrade to `v1.3` to get it working. The implementation works as expected! The only issue I've run into was that the `Node` objects are not deleted after scale down and they're set to a `NotReady` state which leaves a bunch of resources dangling (especially daemonsets).  I don't think this issue is within the scope of this PR, I'm assuming its the responsibility of the API server to delete the node object if the instance is terminating. But I can also see how this problem can become an issue for clusters that run the CA for extended periods of time, wondering if the CA should explicitly tell the API server to delete the node that its terminating as a temporary solution to this problem (again, probably outside the scope of this PR).

**Comment by davidopp** [comment]
Node object should be automatically deleted by node controller once it is gone from the cloud provider:

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/nodecontroller.go#L469

**Comment by osxi** [comment]
Re-tested after adding the AWS SDK autoscaler interface around the `autoscaler` service and still works.

I haven't experienced any dangling nodes. After chatting with @andrewsykim on Slack, it looks like that might be related to the use of EC2 Classic (where my testing took place on a VPC setup).

**Comment by mwielgus** [comment]
```
'golint' problems: 
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/cloudprovider/aws/aws_manager.go:45:6: exported type AutoScaling should have comment or be unexported
```

**Comment by osxi** [comment]
@mwielgus I will fix that

**Comment by osxi** [comment]
@mwielgus can you please manually approve the CLA?

**Comment by mwielgus** [comment]
There is no approval for CLA. There must be something missing on your side. Please check if your git author settings are in line with your github profile.

**Comment by mwielgus** [comment]
The code looks ok. Please squash the commits to 1-2 commits (code / extra godeps) and we should be ready to merge.

**Comment by osxi** [comment]
@mwielgus done

**Comment by mwielgus** [comment]
LGTM

**Comment by mwielgus** [comment]
Thank you for your contribution! :)

**Comment by osxi** [comment]
You're very welcome!

**Comment by pbitty** [comment]
Awesome.  Nice work @osxi and @andrewsykim !

**Comment by mwielgus** [comment]
Please remember to update Kubernetes kube_up.sh scripts for AWS as well :).

**Pull request merged**